module-system/sov-state/src/map.rs

```diff
@@ -21,57 +22,118 @@ pub enum Error {
     MissingValue(Prefix, StorageKey),
 }
 
-impl<K: BorshSerialize, V: BorshSerialize + BorshDeserialize> StateMap<K, V> {
-    pub fn new(prefix: Prefix) -> Self {
+impl<K, V> StateMap<K, V>
+where
+    K: BorshSerialize,
+    V: BorshSerialize + BorshDeserialize,
+{
+    /// Creates a new `StateMap`.
+    pub const fn new(prefix: Prefix) -> Self {
         Self {
             _phantom: (PhantomData, PhantomData),
             prefix,
         }
     }
 
     /// Inserts a key-value pair into the map.
-    pub fn set<S: Storage>(&self, key: &K, value: &V, working_set: &mut WorkingSet<S>) {
-        working_set.set_value(self.prefix(), key, value)
+    pub fn set<S, Q>(&self, key: &Q, value: &V, working_set: &mut WorkingSet<S>)
+    where
+        S: Storage,
+        Q: BorshSerialize + ?Sized,
+        K: Borrow<Q>,
+    {
+        working_set.set_value(self.prefix(), &key, value)
     }
 
     /// Returns the value corresponding to the key or None if key is absent in the StateMap.
-    pub fn get<S: Storage>(&self, key: &K, working_set: &mut WorkingSet<S>) -> Option<V> {
-        working_set.get_value(self.prefix(), key)
+    ///
+    /// # Examples
+    ///
+    /// We can use as argument any type that can be borrowed by the key.
+    ///
+    /// ```rust
+    /// use sov_state::{StateMap, Storage, WorkingSet};
+    ///
+    /// fn foo<S>(map: StateMap<Vec<u8>, u64>, key: &[u8], ws: &mut WorkingSet<S>) -> Option<u64>
+    /// where
+    ///     S: Storage,
+    /// {
+    ///     // we perform the `get` with a slice, and not the `Vec`. it is so because `Vec` borrows
+    ///     // `[T]`.
+    ///     map.get(key, ws)
+    /// }
+    /// ```
+    ///
+    /// However, some concrete types won't implement `Borrow`, but we can easily cast them into
+    /// common types that will
+    ///
+    /// ```rust
+    /// use sov_state::{StateMap, Storage, WorkingSet};
+    ///
+    /// fn foo<S>(map: StateMap<Vec<u8>, u64>, key: [u8; 32], ws: &mut WorkingSet<S>) -> Option<u64>
+    /// where
+    ///     S: Storage,
+    /// {
+    ///     map.get(&key[..], ws)
+    /// }
+    /// ```
```

Comment:
Very good point. I don't have anything to add at this very moment, I'll go back to my PR and see if there's a way to reconcile these two constraints.

Reply:
> We are adding another requirement here: serialized version of borrowed and owned value must be equivalent.

Yes, by adding this requirement we allow to reduce cloning. 
Question is, can we express this restriction in the code, so users won't get hit by 

> Is it even true for borsh?

But what if you have ZST as key in the first place? Would it suppose to break?

So I have this test and it looks like this;

```rust
   use sov_state::{DefaultStorageSpec, Prefix, ProverStorage, StateMap};
    use super::*;

    #[test]
    fn check_something() {
        let tmpdir = tempfile::tempdir().unwrap();
        let mut working_set =
            WorkingSet::new(ProverStorage::<DefaultStorageSpec>::with_path(tmpdir.path()).unwrap());

        let prefix = Prefix::new("test".as_bytes().to_vec());
        let map_1: StateMap<Vec<()>, u8> = StateMap::new(prefix.clone());

        let d1 = [(); 2 ^ 64].to_vec();

        map_1.set(&d1, &3, &mut working_set);
        let a = map_1.get(&d1, &mut working_set);
        println!("A: {:?}", a);

        let map_2: StateMap<&'static [()], u8> = StateMap::new(prefix);

        let d2 = [(); 2 ^ 64];
        map_2.set(&&d2[..], &3, &mut working_set);

        let b = map_2.get(&&d2[..], &mut working_set);
        println!("B: {:?}", b);

        assert!(a.is_some());
        assert!(b.is_some());
        assert_eq!(a, b);
    }
```

Nothing blew up. But probably I misunderstood question.



I hope we can agree that there's usability problem. Often you have only reference to slice of bytes and you need to use it as key, so you have to clone it. Then during set it is cloned again:

```rust
impl StorageKey {
    pub fn new<K: BorshSerialize>(prefix: &Prefix, key: &K) -> Self {
        let encoded_key = key.try_to_vec().unwrap();
```
        
So we trade off API purity for double cloning. We need to understand that. 
And in current scenario, we push burden to our customers. So I think it is in our interest to find solution to this problem. Maybe we need to add some constraints to `codec` type, so it won't be easy to use coded which has different serialization. Maybe each codec have to implement those kind of checks for types that can be keys.

Reply:
I know it is not strong reason, but how likely it is that someone will use ZST for keys?

Also, can we limit CacheKey current implementation not to all AsRef<T> but only to Vec<u8> and slice of bytes. That might be option, since that will be most probable use case.

Reply:
So the API breaks:
```
map.get(&vec[()]) 
```
panics while 
```
map.get(&[();1]);
``` 
won't. 

So it is inconsistent. 
The ZST and slice vs vec is just an example of a bigger problem. We add another requirement to `Borrow`, and all the serialization formats in cartes.io don't care about it. So to use it safely a user has to examine all types which can be borrowed and check if the implementation of the sterilization format they choose doesn't break the assumption (and won't break it in the future).  We already saw that it is a broken event in the simplest case. 

We could have another trait `OurBorrow` with the additional requirement documented but still, 3rd party serialization crates won't care.

Reply:
@citizen-stig your example gave me an idea. You tested `Vec<()>` against `&[()]`, which have the same encodings in Borsh, to it succeeds. What about arrays, though? Arrays don't have a length marker because there's no need for those, yet they `Borrow` to slices. This test fails:

```rust
#[test]
fn check_something() {
    use super::*;
    use crate::{DefaultStorageSpec, Prefix, ProverStorage};

    let tmpdir = tempfile::tempdir().unwrap();
    let mut working_set =
        WorkingSet::new(ProverStorage::<DefaultStorageSpec>::with_path(tmpdir.path()).unwrap());

    let prefix = Prefix::new("test".as_bytes().to_vec());

    type Key = [u8; 4];
    let map: StateMap<Key, u32> = StateMap::new(prefix.clone());

    let key = [0u8; 4];
    let key_ref_array: &[u8; 4] = <Key as Borrow<[u8; 4]>>::borrow(&key);
    let key_ref_slice: &[u8] = <Key as Borrow<[u8]>>::borrow(&key);

    // We set with &[u8; 4]...
    map.set(key_ref_array, &3, &mut working_set);
    // ...getting with &[u8; 4] works, obviously...
    assert!(map.get(key_ref_array, &mut working_set).is_some());
    // ...but changing the target of the Borrow to &[u8] doesn't.
    assert!(map.get(key_ref_slice, &mut working_set).is_some());
}
```

Reply:
@bkolad I didn't see your comment before posting mine, but yeah basically +1.

Reply:
One possibility would be to extend `StateKeyCodec` and shift the responsibility of ensuring that `K` and `Q` serialize to the same byte sequence to the codec author, instead of the caller.

```rust
pub trait StateKeyCodec<K, Q = &K> {
    type KeyError: std::fmt::Debug;

    fn encode_key(&self, key: &Q) -> Vec<u8>;

    fn try_decode_key(&self, bytes: &[u8]) -> Result<K, Self::KeyError>;

    // ...
}
```

Effectively, the default type parameter `Q = K` would be an opt-in mechanism. We'd have to modify this PR to allow *only* the types `Q` for which the codec is valid: `impl<K, V, C, Q> StateMap<K, V, C> where C: StateKeyCodec<K, Q> + StateValueCodec<V>`.

Blanket implementations like the following won't support the borrowee pattern, and it'll be up to us to choose when and for which types to add support:

```rust
impl<K> StateKeyCodec<K> for BorshCodec
where
    K: borsh::BorshSerialize + borsh::BorshDeserialize,
{
   // ...
}
```

I'd have to think it more through, but I think it might work. Whether or not it's worth it is another story.

Reply:
@bkolad 

So

```rust
map.get(&vec[()]) 

// panics while
```
Does not panic for me:

```rust
        type Key = Vec<()>;
        let map: StateMap<Key, u32> = StateMap::new(prefix.clone());

        let a = map.get(&vec![()], &mut working_set);
        assert!(a.is_none())
```


@neysofu , thank you for your example, I can confirm that it does not fetch key correctly =(.

I just want to reiterate my point. I am not trying to prove that ZST types suppose to work. Or that serialization adds extra constraints.

I just want to empathize, that there's undeniably usability problem for people in a simple case of `&Vec<u8>` or `&String`. And we have 2 options:

1. "Sorry devs, that's life, deal with it". 
2. Try to find some practical solution, that will improve usability and bring value to the users. 

In my opinion second option is preferable, even if it is harder. 
Yes, people can brake HashMap, and that's fine, because everyday usage is more important, than edge cases. 
I think the same way if we push for this convention, it can be easier to use our SDK everyday.


> So to use it safely a user has to examine all types which can be borrowed and check if the implementation of the sterilization format they choose doesn't break the assumption (and won't break it in the future). We already saw that it is a broken event in the simplest case.

Can you please provide another example, besides ZST, which can brake it? 

>  all types which can be borrowed

All types that are used as keys. 

> (and won't break it in the future)

Breaking change in serialization library is a big thing on itself. If you have running system and serialization for Vec<T> has changed, you cannot just upgrade.


Anyway.

Other options that we can use:

- some automatic way to check code base, that anything used in a key, should have same serialization with given coded. We can even add macro that will generate those tests.
- Update our documentation to make sure that if people select something special for the case, that they should be carefule.

Reply:
@citizen-stig we are on the old version of borsh, the check and panic is in the latest `master`. 
I am not against improving it if we can I just don't see any way. `

> Can you please provide another example, besides ZST, which can brake it?

There are infinite amount  of ways of braking it. If I write a format which seializes `Foo` differently than `Bar` and `Foo` can be borrowed as `Bar` this is enough and I didn't make any mistake. For `Eq` it would be an error in my code not an edge case. If @neysofu solution works and we don't complicate the code much then great but as for now I see the problem as fundamental security bridge.

Reply:
I'll try putting something together :+1: